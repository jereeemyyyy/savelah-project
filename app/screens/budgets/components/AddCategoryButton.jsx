import React from 'react';
import { TouchableOpacity, View, Text } from 'react-native';
import { FontAwesome6 } from '@expo/vector-icons';

const AddCategoryButton = ({ onPress }) => {
  return (
    <TouchableOpacity onPress={onPress}>
      <View className="bg-purple-500 w-30 h-10 rounded-lg p-2 items-center flex-row">
        <FontAwesome6 name="add" size={18} color="white" />
        <Text className="text-white text-xs font-bold ml-2 ">Add Category</Text>
      </View>
    </TouchableOpacity>
  );
};

export default AddCategoryButton;