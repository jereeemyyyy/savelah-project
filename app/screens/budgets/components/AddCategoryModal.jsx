import React, { useState } from 'react';
import { View, Modal, TextInput, TouchableOpacity, Text, Alert } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import ChooseIconPopup from './ChooseIconPopup';
import { MaterialIcons } from '@expo/vector-icons';

export default function AddCategoryModal({ visible, onClose, onAddCategory }) {
  const [categoryName, setCategoryName] = useState('');
  const [categoryAmount, setCategoryAmount] = useState('');
  const [categoryIcon, setCategoryIcon] = useState(null);
  const [showIconPopup, setShowIconPopup] = useState(false);

  const handleIconSelect = (iconName) => {
    setCategoryIcon(iconName);
    setShowIconPopup(false);
  };

  // const handleAddCategory = () => {
  //   if (categoryName && categoryAmount && categoryIcon) {
  //     const newCategory = {
  //       name: categoryName,
  //       amount: parseFloat(categoryAmount),
  //       icon: categoryIcon,
  //     };
  //     onAddCategory(newCategory);
  //     setCategoryName('');
  //     setCategoryAmount('');
  //     setCategoryIcon(null);
  //     onClose();
  //   }
  // };

  const handleAddCategory = () => {
    const missingFields = [];

    if (!categoryName.trim()) missingFields.push('Category Name');
    if (!categoryAmount.trim()) missingFields.push('Category Amount');
    if (!categoryIcon) missingFields.push('Category Icon');

    if (missingFields.length > 0) {
      Alert.alert(
        'Incomplete Information',
        `Please fill in the following field(s): ${missingFields.join(', ')}`,
        [{ text: 'OK' }]
      );
    } else {
      const newCategory = {
        name: categoryName.trim(),
        amount: parseFloat(categoryAmount),
        icon: categoryIcon,
      };
      onAddCategory(newCategory);
      setCategoryName('');
      setCategoryAmount('');
      setCategoryIcon(null);
      onClose();
    }
  };

  return (
    <Modal visible={visible} animationType="fade" transparent>
      <View className="flex-1 bg-black/50">
        <View className="flex-1 justify-center items-center">
          <View className="bg-white p-6 rounded-lg justify-center p-3">
            <View className="flex-row justify-between items-center mb-3">  
                <Text className="text-black font-bold text-2xl ml-9">Add Category</Text>
                <TouchableOpacity onPress={onClose} className="mr-2">
                    <MaterialIcons name="cancel" size={24} color="black"/> 
                </TouchableOpacity>
            </View>
            
            <TextInput
              placeholder="Category Name"
              placeholderTextColor="gray"
              value={categoryName}
              onChangeText={setCategoryName}
              style={{ color: 'black' }}
              className="border border-gray-400 p-2 mb-3 w-60"
            />
            <TextInput
              placeholder="Category Amount"
              placeholderTextColor="gray"
              value={categoryAmount}
              onChangeText={setCategoryAmount}
              style={{ color: 'black' }}
              keyboardType="numeric"
              className="border border-gray-400 p-2 mb-3"
            />
            <View className="items-center"> 
                <TouchableOpacity
                onPress={() => setShowIconPopup(true)}
                className="flex-row items-center mb-2"
                >
                {categoryIcon ? (
                    <Ionicons name={categoryIcon} size={24} color="black" />
                ) : (
                    <View className="bg-indigo-500 rounded p-2">
                        <Text className="text-white font-bold">Choose Icon</Text>
                    </View>
                    
                )}
                </TouchableOpacity>
            </View>
            <View className="items-center">
                <TouchableOpacity
                onPress={handleAddCategory}
                className="bg-blue-500 py-2 px-9 rounded bg-violet-500 justify-center items-center"
                >
                <Text className="text-white font-bold">Add</Text>
                </TouchableOpacity>
            </View>
          </View>
        </View>
      </View>
      <ChooseIconPopup
        visible={showIconPopup}
        onIconSelect={handleIconSelect}
        onClose={() => setShowIconPopup(false)}
      />
    </Modal>
  );
}