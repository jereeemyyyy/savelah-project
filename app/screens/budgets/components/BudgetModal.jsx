import React from 'react';
import { View, Text, TouchableOpacity, Modal, Dimensions, Alert} from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import Slider from '@react-native-community/slider';
import { Entypo } from '@expo/vector-icons';

const { height } = Dimensions.get('window');

const BudgetModal = ({ visible, onClose, budgetGoal, totalBudget, onBudgetChange, onSave }) => {
  
  // Settings for the modal dimensions
  const modalHeight = height / 2;
  const topSectionHeight = 2 * modalHeight / 6;
  const bottomSectionHeight = modalHeight - topSectionHeight;

  // Creates the alert box
  const createTwoButtonAlert = () =>
    Alert.alert('Confirmation', `Are you sure you want to add $${budgetGoal} into this category?` , [
      {
        text: 'Cancel',
        onPress: () => console.log('Cancel Pressed'),
        style: 'cancel',
      },
      {text: 'OK', onPress: () => console.log('OK Pressed')},
    ]);

    const handleSave = () => {
      // Handle Query logic here to save the budget to supabase

    }

  return (
    <Modal visible={visible} animationType="slide" transparent>
      <View
        className="absolute bottom-0 left-0 right-0 rounded-t-lg"
        style={{ height: modalHeight }}
      >
        <View className="bg-white rounded-t-lg" style={{ height: topSectionHeight }}>
          <View className="flex-1 flex-col justify-end p-2 mt-3 mb-7">
            <TouchableOpacity onPress={onClose}
                              className="p-2 content-start"
            >
                <Ionicons name="arrow-back-circle" size={28} color="black" />
            </TouchableOpacity>
            <View className="flex-row justify-between">
              <Text className="ml-10  font-bold text-black text-5xl">${budgetGoal}</Text>
              <View className="mt-1 mr-2 flex-row">
                <TouchableOpacity className="mr-3">
                  <Ionicons name="calculator" size={34} color="black" />
                </TouchableOpacity>
                <TouchableOpacity>
                    <Entypo name="circle-with-cross" size={34} color="black" />
                </TouchableOpacity>
              </View>
            </View>
          </View>
        </View>
        <View className="bg-violet-500 rounded-xl" style={{ height: bottomSectionHeight }}>
          <View className="h-full justify-between p-6">
            <View className="">
              <Slider
                style={{ width: '100%', height: 40 }}
                minimumValue={0}
                maximumValue={totalBudget}
                minimumTrackTintColor="#FFFFFF"
                maximumTrackTintColor="#000000"
                thumbTintColor="#FFFFFF"
                value={budgetGoal}
                onValueChange={onBudgetChange}
                step={1}
              />
              <Text className="text-center text-white mt-2">${budgetGoal.toFixed(0)}</Text>
            </View>
            <View className="flex-col flex-1 justify-start">
                <Text className="text-white px-2 font-bold text-xl">
                {budgetGoal > totalBudget ? 'Over Budget' : 'Normal'}
                </Text>
                <Text className="my-2 ml-2 text-white">
                    {budgetGoal > totalBudget 
                     ? 'Oh no! You exceeded your budget' 
                     : 'Its ok! This amount is just within your budget!'}
                </Text>
            </View> 
            <TouchableOpacity className="bg-white rounded-full py-3 mt-4" onPress={createTwoButtonAlert}>
              <Text className="text-center text-purple-500 font-bold">Save</Text>
            </TouchableOpacity>
          </View>
        </View>
      </View> 
    </Modal>
  );
};

export default BudgetModal;
