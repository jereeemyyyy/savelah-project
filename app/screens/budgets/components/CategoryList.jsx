import React from 'react';
import { ScrollView, View, Text} from 'react-native';
import CategoryItem from './CategoryItem'

const CategoryList = ({ categories, pressed, selectedCategory, handleCategorySelect }) => {
 
  
  return (
    <View className="flex-1">
        {categories.length <= 0 ? (
        <View className="flex-1 justify-center items-center">
          <Text className="text-white text-lg">No Categories available.</Text>
          <Text className="text-white text-m mb-8">Create one by clicking the Add Category button!</Text>
        </View>
      ) : (
        <ScrollView
          horizontal
          showsHorizontalScrollIndicator={false}
          className="ml-12 mb-6"
        >
          {categories.map((category) => (
            <CategoryItem
              key={category.id}
              category={category}
              pressed={pressed}
              selectedCategory={selectedCategory}
              handleCategorySelect={handleCategorySelect}
            />
          ))}
        </ScrollView>  
      )}
    </View>
    
  );
};

export default CategoryList;