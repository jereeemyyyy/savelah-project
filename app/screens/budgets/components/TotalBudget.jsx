import React from 'react';
import { Text, View } from 'react-native';

const TotalBudget = ({ totalBudget }) => {
  return (
    <View className="flex-row">
      <Text className="text-white">Total budget is </Text> 
      <Text className="text-white font-bold">${totalBudget}</Text>
    </View>
    
  );
};

export default TotalBudget;