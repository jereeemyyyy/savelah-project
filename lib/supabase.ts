import { AppState } from 'react-native'
import 'react-native-url-polyfill/auto'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = 'https://cjcldaiiqmedjmojjzfz.supabase.co'
const supabaseAnonKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNqY2xkYWlpcW1lZGptb2pqemZ6Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MTY3OTE2NDksImV4cCI6MjAzMjM2NzY0OX0.6z2M2mG_qmSPPwujMWYpN4KkLgiU0mWjGdsCdVZvlNw'

export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    storage: AsyncStorage,
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: false,
  },
})

// Tells Supabase Auth to continuously refresh the session automatically
// if the app is in the foreground. When this is added, you will continue
// to receive `onAuthStateChange` events with the `TOKEN_REFRESHED` or
// `SIGNED_OUT` event if the user's session is terminated. This should
// only be registered once.
AppState.addEventListener('change', (state) => {
  if (state === 'active') {
    supabase.auth.startAutoRefresh()
  } else {
    supabase.auth.stopAutoRefresh()
  }
})