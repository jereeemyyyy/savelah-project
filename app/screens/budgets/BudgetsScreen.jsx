import React, { useState, useEffect } from 'react';
import { View, Text, Alert } from 'react-native';
import { supabase } from '../../../lib/supabase';
import BudgetModal from './components/BudgetModal';
import AddCategoryModal from './components/AddCategoryModal';
import TotalBudget from './components/TotalBudget';
import CategoryList from './components/CategoryList';
import AddCategoryButton from './components/AddCategoryButton';


export default function BudgetsScreen() {
  const [budgetGoal, setBudgetGoal] = useState(400);
  const [totalBudget, setTotalBudget] = useState(0);
  const [selectedCategory, setSelectedCategory] = useState(null);
  const [showModal, setShowModal] = useState(false);
  const [pressed, setPressed] = useState(null);
  const [showAddCategoryModal, setShowAddCategoryModal] = useState(false);
  const [categories, setCategories] = useState([]);
  const [userId, setUserId] = useState(null);

  useEffect(() => {
    fetchUserCategories(); 
    categoriesListener();
  }, []);

  const categoriesListener = async () => {
    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (user) {
        const subscription = supabase
        .channel('room1')
        .on('postgres_changes', 
          { event: '*', 
            schema: 'public', 
            table: 'categories', 
            filter: `user_id=eq.${user.id}` 
          }, payload => {
            console.log('Change received!', payload);
            fetchUserCategories(); 
          })
          .subscribe();
          
        return () => {
          fetchUserCategories();
        };
      }
    } catch (error) {
        console.error('Error fetching categories:', error.message);
    }
   };   

  const fetchUserCategories = async () => {
    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (user) {
        const { data, error } = await supabase.rpc('get_user_categories', { p_user_id: user.id });
        if (error) throw error;
        setCategories(data || []);
        updateTotalBudget(data);
        
      } else {
        console.log("No user found");
      }
    } catch (error) {
      console.error('Error fetching categories:', error.message);
    }
    
  };

  const updateTotalBudget = (categories) => {
    const total = categories.reduce((sum, category) => sum + category.amount, 0);
    setTotalBudget(total);
  };


  const handleCategorySelect = (category) => {
    setSelectedCategory(category);
    setShowModal(true);
    setPressed(category.id);
    setBudgetGoal(category.amount);
  };

  const handleBudgetChange = (value) => {
    setBudgetGoal(value);
  };

  const handleCloseModal = () => {
    setShowModal(false);
    setPressed(null);
  }

  // Save the new budget goal (Update)
  const handleSaveNewBudget = async () => {
    try {
      const { data, error } = await supabase
        .from('categories')
        .update({ amount: budgetGoal })
        .eq('id', selectedCategory.id);

      if (error) {
        throw new Error('Error updating category');
      }

      setCategories(categories.map(category =>
        category.id === selectedCategory.id ? { ...category, amount: budgetGoal } : category
      ));
    } catch (error) {
      console.error('Error updating category:', error.message);
    }

    setShowModal(false);
  };


  // Deletes a category
  const handleDeleteCategory = async () => {
    try {
      const { data: expenses, error: fetchExpensesError } = await supabase
        .from('expenses')
        .select('*')
        .eq('category_id', selectedCategory.id);
  
      if (fetchExpensesError) {
        console.error('Error fetching expenses:', fetchExpensesError.message);
        return;
      }
  
      if (expenses.length > 0) {
        Alert.alert(
          'Delete Category',
          'This category has associated expenses. Are you sure you want to delete this category and all its associated expenses?',
          [
            {
              text: 'Cancel',
              onPress: () => console.log('Cancel Pressed'),
              style: 'cancel',
            },
            {
              text: 'Delete',
              onPress: async () => {
                try {
                  const { error } = await supabase.rpc('delete_category_and_expenses', {
                    p_category_id: selectedCategory.id
                  });
  
                  if (error) {
                    console.error('Error deleting category and expenses:', error.message);
                    return;
                  }
  
                  setCategories(categories.filter((category) => category.id !== selectedCategory.id));
                  setShowModal(false);
                } catch (error) {
                  console.error('Error deleting category and expenses:', error.message);
                }
              },
            },
          ]
        );
      } else {
        const { error } = await supabase.rpc('delete_category_and_expenses', {
          p_category_id: selectedCategory.id
        });
  
        if (error) {
          console.error('Error deleting category:', error.message);
        } else {
          setCategories(categories.filter((category) => category.id !== selectedCategory.id));
          setShowModal(false);
        }
      }
    } catch (error) {
      console.error('Error deleting category:', error.message);
    }
  };

  // Add a new category function
  const handleAddCategory = async (newCategory) => {

    try {
      const { data: { user }, error: userError } = await supabase.auth.getUser();
  
      if (userError) {
        throw new Error('Error fetching user');
      }
  
      if (!user) throw new Error('User not authenticated');
  
      
      // Capitalize the first letter of newCategory.name
      newCategory.name = newCategory.name.charAt(0).toUpperCase() + newCategory.name.slice(1).toLowerCase();
      
      // Check if the category already exists for the user
      const { data: existingCategories, error: fetchError } = await supabase
        .from('categories')
        .select('category')
        .eq('user_id', user.id) 
        .eq('category', newCategory.name);
  
      if (fetchError) {
        console.log(fetchError);
        throw new Error(fetchError);
      }
  
      if (existingCategories && existingCategories.length > 0) {
        alert('This category has already been created');
        return;
      }
  
      const { data, error } = await supabase
        .from('categories')
        .insert({ 
          amount: newCategory.amount, 
          icon: newCategory.icon, 
          category: newCategory.name 
        })
        .select();
  
      if (error) {
        console.log('Error adding category:', error);
      } else {
        setCategories([...categories, ...data]);
        setShowAddCategoryModal(false);
      }
    } catch (error) {
      console.log('Error adding category:', error);
    }
    fetchUserCategories();
  };
  

  return (
    <View className="flex-1 bg-gray-800 p-6">
      <Text className="text-white text-4xl font-bold mt-3 mb-6">
        Set up a monthly budget goal
      </Text>

      <View className="flex-row justify-between mb-6 items-center">
        <TotalBudget totalBudget={totalBudget} />
        <AddCategoryButton onPress={() => setShowAddCategoryModal(true)} />
      </View>
      
    
      <CategoryList 
        categories={categories} 
        pressed={pressed} 
        selectedCategory={selectedCategory}
        handleCategorySelect={handleCategorySelect} 
      />
      
      
      <BudgetModal
      visible={showModal}
      onClose={handleCloseModal}
      budgetGoal={budgetGoal}
      totalBudget={totalBudget}
      onBudgetChange={handleBudgetChange}
      onSave={handleSaveNewBudget}
      onDelete={handleDeleteCategory}
      />

      <AddCategoryModal
        visible={showAddCategoryModal}
        onClose={() => setShowAddCategoryModal(false)}
        onAddCategory={handleAddCategory}
      />
    </View>
  );
}