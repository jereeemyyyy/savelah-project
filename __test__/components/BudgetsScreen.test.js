import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { View, Text } from 'react-native'; // Import necessary components from react-native
import BudgetsScreen from '../../app/screens/budgets/BudgetsScreen';
import { supabase } from '../../lib/supabase';
import BudgetModal from '../../app/screens/budgets/components/BudgetModal';
import AddCategoryModal from '../../app/screens/budgets/components/AddCategoryModal';
import TotalBudget from '../../app/screens/budgets/components/TotalBudget';
import CategoryList from '../../app/screens/budgets/components/CategoryList';
import AddCategoryButton from '../../app/screens/budgets/components/AddCategoryButton';

jest.mock('../../lib/supabase', () => ({
    supabase: {
        auth: {
            getUser: jest.fn(),
        },
        rpc: jest.fn(),
        channel: jest.fn().mockReturnValue({
            on: jest.fn().mockReturnThis(),
            subscribe: jest.fn().mockResolvedValue({
                data: { categories: [] },
            }),
        }),
    },
}));


jest.mock('../../app/screens/budgets/components/BudgetModal', () => 'BudgetModal');
jest.mock('../../app/screens/budgets/components/AddCategoryModal', () => 'AddCategoryModal');
jest.mock('../../app/screens/budgets/components/TotalBudget', () => 'TotalBudget');
jest.mock('../../app/screens/budgets/components/CategoryList', () => 'CategoryList');
jest.mock('../../app/screens/budgets/components/AddCategoryButton', () => 'AddCategoryButton');

describe('BudgetsScreen', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });
        supabase.rpc.mockResolvedValue({ data: [] });

        let tree;
        await act(async () => {
            tree = renderer.create(<BudgetsScreen />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('contains the expected text elements', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });
        supabase.rpc.mockResolvedValue({ data: [] });

        let component;
        await act(async () => {
            component = renderer.create(<BudgetsScreen />);
        });
        const instance = component.root;

        expect(instance.findByType(Text).props.children).toContain('Set up a monthly budget goal');
    });

    it('contains the BudgetModal, AddCategoryModal, TotalBudget, CategoryList, and AddCategoryButton components', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });
        supabase.rpc.mockResolvedValue({ data: [] });

        let component;
        await act(async () => {
            component = renderer.create(<BudgetsScreen />);
        });
        const instance = component.root;

        expect(instance.findByType('BudgetModal')).toBeTruthy();
        expect(instance.findByType('AddCategoryModal')).toBeTruthy();
        expect(instance.findByType('TotalBudget')).toBeTruthy();
        expect(instance.findByType('CategoryList')).toBeTruthy();
        expect(instance.findByType('AddCategoryButton')).toBeTruthy();
    });
});
