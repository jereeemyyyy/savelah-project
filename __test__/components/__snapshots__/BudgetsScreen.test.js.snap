// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetsScreen renders correctly 1`] = `
<View
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "backgroundColor": "#1f2937",
      },
      {
        "paddingBottom": 24,
        "paddingLeft": 24,
        "paddingRight": 24,
        "paddingTop": 24,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#fff",
        },
        {
          "fontSize": 36,
          "lineHeight": 40,
        },
        {
          "fontWeight": "700",
        },
        {
          "marginTop": 12,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    Set up a monthly budget goal
  </Text>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "justifyContent": "space-between",
        },
        {
          "marginBottom": 24,
        },
        {
          "alignItems": "center",
        },
      ]
    }
  >
    <TotalBudget
      totalBudget={0}
    />
    <AddCategoryButton
      onPress={[Function]}
    />
  </View>
  <CategoryList
    categories={[]}
    handleCategorySelect={[Function]}
    pressed={null}
    selectedCategory={null}
  />
  <BudgetModal
    budgetGoal={400}
    onBudgetChange={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    totalBudget={0}
    visible={false}
  />
  <AddCategoryModal
    onAddCategory={[Function]}
    onClose={[Function]}
    visible={false}
  />
</View>
`;
